docs/proposal/20220712-garbage-collection.md

```diff
@@ -0,0 +1,338 @@
+
+### Goals
+
+1. To delete AWS resources that where created by CCM in the child/tenant cluster.
+2. To work across unmanaged (i.e. EC2 control plane) and managed (i.e. EKS) clusters.
+3. Solution must work in a scenario where GitOps is used.
+
+### Non-Goals/Future Work
+
+- Delete EBS volumes created by the CCM
+  - This will be considered as part of future work
+- Clean-up other resources created by something other than the CCM (for example a custom operator)
+- Fine grained control of which clusters will be garbage collected or not
+  - Initially if the feature is enabled it will be an opt-out model
+  - We will add finer grained control in a future enhancement
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an platform operator/engineer
+I want to delete a cluster and all its associated AWS resources
+When not using GitOps
+So that there are no orphaned/unused AWS resources
+
+#### Story 2
+
+As a platform operator/engineer
+I want to be able to delete a cluster and all its associated AWS resources 
+When using a GitOps tools (like Flux/Argo)
+So that there are no orphaned/unused AWS resources
+
+#### Story 3
+
+As a platform operator/engineer
+I want to be able to opt-out a cluster of being garbage collected
+
+#### Story 4
+
+As a platform operator/engineer
+I want to be able to opt-in/opt-out a cluster for garbage collection
+After it has been created
+So that i can investigate/overcome issues
+
+## Requirements
+
+### Functional
+
+<a name="FR1">FR1.</a> CAPA MUST support cleaning up of AWS resources created by the CCM for a tenant cluster when not using GitOps.
+
+<a name="FR2">FR2.</a> CAPA MUST support cleaning up of AWS resources created by the CCM for a tenant cluster when using GitOps.
+
+<a name="FR3">FR3.</a> CAPA MUST support cleaning up of AWS resources for unmanaged and managed clusters
+
+<a name="FR4">FR4.</a> CAPA MUST support a way to opt-out of garbage collection at the point of cluster creation.
+
+<a name="FR5">FR5.</a> CAPI MUST not allow me to delete a cluster fullt until garbage collection has occurred.
+
+<a name="FR6">FR6.</a> CAPA SHOULD provide a way for me to opt-in or opt-out a cluster from garbage collection AFTER it has been created.
+
+### Non-Functional
+
+<a name="NFR8">NFR8.</a> CAPA MUST be able to easily add additional AWS resource cleanup in the future.
+
+<a name="NFR9">NFR9.</a> Unit tests MUST exist for new garbage collection code.
+
+<a name="NFR10">NFR10.</a> e2e tests MUST exist for the new garbage collection code for both unmanaged and managed clusters..
+
+### Implementation Details/Notes/Constraints
+
+#### Proposed Changes
+
+In the initial implementation of garbage collection if the feature is enabled all clusters will by default be marked for garbage collection. But we will supply a means to opt-out at time of creation as per [FR4](#FR4).
+
+The proposed changes will be documented depending on if the cluster is being created/updated or deleted.
```

Comment:
```suggestion
The proposed changes will be documented depending on if the cluster is being created, updated or deleted.
```

This sentence is a bit confusing. The documentation will be based on what the cluster action is? Or each action will have its own document? :D Consider rewriting this to make it a bit clearer, please. :)

Reply:
Reworded.